### PR TITLE
 Backport unicode litterals from Python 3

### PR DESCRIPTION
**Pull request by fpagnoux**
Minor Change without impacts for reusers:
  - Make code more Python3-like by backporting unicode litterals.
  - With this backport, all strings are by default unicodes.
  - The `u` prefix for strings should *not* be used anymore.
  - Each new module must start by `from __future__ import unicode_literals` for the backport to be effective.

### Rationale

We've paid the price for compatibility with Python 3, let's start to enjoy some of its  benefits 🙂.

The general idea is to:
- Opportunistically apply Python-3 standards to our codebase
- Rely on `__future__` (Python builtins backports) and `future` (extra library) to keep compatibility with Python 2
- When we drop compatibility with Python 2, we'll mostly just have  imports to remove

See this talk: https://www.youtube.com/watch?v=klaGx9Q_SOA

### CODE REVIEW

**Comment by MattiSG** [comment]
OMG so much yes!!! Thanks!! ❤️

**Comment by fpagnoux** [comment]
This PR doesn't contain much logic, and should be relatively quick to review, but has a _huge_ potential for conflicts, so a quick review would be really appreciated 🙂.

@maukoquiroga @sandcha @Anna-Livia

File: openfisca_web_api_preview/loader/tax_benefit_system.py [hunk 1]
```diff
@@ -13,9 +14,9 @@ def build_tax_benefit_system(country_package_name):
         country_package = importlib.import_module(country_package_name)
     except ImportError:
         message = linesep.join([traceback.format_exc(),
-                                u'Could not import module `{}`.'.format(country_package_name).encode('utf-8'),
```
Comment (magopian) [inline openfisca_web_api_preview/loader/tax_benefit_system.py]: Wow, weird, how was it working previously? It was joining one line of bytes with lines of str? How was that possible? I must be missing something...

File: openfisca_web_api_preview/loader/variables.py [hunk 2]
```diff
@@ -38,9 +39,9 @@ def build_source_url(country_package_metadata, source_file_path, start_line_numb
 
 def build_formula(formula, country_package_metadata, source_file_path, tax_benefit_system):
     source_code, start_line_number = inspect.getsourcelines(formula)
-    if source_code[0].lstrip(' ').startswith('@'):  # remove decorator
```
Comment (magopian) [inline openfisca_web_api_preview/loader/variables.py]: did you remove the decorator removal on purpose?

File: tests/core/test_countries.py [hunk 3]
```diff
@@ -159,7 +160,7 @@ def test_calculate_variable_with_wrong_definition_period():
     expected_words = ['period', '2016', 'month', 'basic_income', 'ADD']
 
     for word in expected_words:
-        assert word in error_message, u'Expected "{}" in error message "{}"'.format(word, error_message).encode('utf-8')
+        assert word in error_message, 'Expected "{}" in error message "{}"'.format(word, error_message).encode('utf-8')
```
Comment (magopian) [inline tests/core/test_countries.py]: what is the `.encode('utf-8')` for here?

File: tests/core/test_yaml.py [hunk 4]
```diff
@@ -17,11 +18,6 @@
 openfisca_core_dir = pkg_resources.get_distribution('OpenFisca-Core').location
 yaml_tests_dir = os.path.join(openfisca_core_dir, 'tests', 'core', 'yaml_tests')
 
-# When tests are run by the CI, openfisca_core is not installed in editable mode, therefore the tests are not packaged with the ditribution. We thus get them from ~/openfisca-core/tests/core/yaml_tests
-if not os.path.isdir(yaml_tests_dir):
-    ci_yaml_tests_dir = os.path.join(os.path.expanduser('~'), 'openfisca-core', 'tests', 'core', 'yaml_tests')
-    yaml_tests_dir = ci_yaml_tests_dir if os.path.isdir(ci_yaml_tests_dir) else yaml_tests_dir
-
```
Comment (magopian) [inline tests/core/test_yaml.py]: That doesn't seem to be related to the issue, am I missing something?

File: tests/web_api/basic_case/test_variables.py [hunk 5]
```diff
@@ -13,7 +14,7 @@ def assert_items_equal(x, y):
 # /variables
 
 variables_response = subject.get('/variables')
-GITHUB_URL_REGEX = '^https://github\.com/openfisca/openfisca-country-template/blob/\d+\.\d+\.\d+((.dev|rc)\d+)?/openfisca_country_template/variables/(.)+\.py#L\d+-L\d+$'
+GITHUB_URL_REGEX = '^https://github\.com/openfisca/country-template/blob/\d+\.\d+\.\d+((.dev|rc)\d+)?/openfisca_country_template/variables/(.)+\.py#L\d+-L\d+$'
```
Comment (magopian) [inline tests/web_api/basic_case/test_variables.py]: That doesn't seem to be related to the issue at hand either

File: openfisca_core/scripts/measure_numpy_condition_notations.py [hunk 6]
```diff
@@ -10,6 +10,8 @@
 
 The aim of this script is to compare the time taken by the calculation of the values
 """
+from __future__ import print_function
```
Comment (Anna-Livia) [inline openfisca_core/scripts/measure_numpy_condition_notations.py]: What features is `print_function`bringing in this file ?

Comment (Anna-Livia) [reply openfisca_core/scripts/measure_numpy_condition_notations.py]: Could we write :
```py
from __future__  import unicode_literals, print_function
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,14 @@
 # Changelog
 
+
+### 23.3.2 [#702](https://github.com/openfisca/openfisca-core/pull/702)
+
+Minor Change without impacts for reusers:
```
[the remark this answers is not in the document]
Comment (fpagnoux) [reply CHANGELOG.md]: I meant re-users, but users is probably better.

File: README.md [hunk 8]
```diff
@@ -6,7 +6,9 @@ This package contains the core features of OpenFisca, which are meant to be used
 
 ## Environment
 
-This package requires Python 2.7
+OpenFisca runs on Python 3.6, or more recent versions.
```
[the remark this answers is not in the document]
Comment (fpagnoux) [reply README.md]: I know! But they are not backportable to 2.7 😞 , unless we use some kind of [transpilation](https://github.com/asottile/future-fstrings), and I'm not sure I want to go there...

File: openfisca_core/scripts/measure_numpy_condition_notations.py (hunk 6 above)
Comment (fpagnoux) [reply openfisca_core/scripts/measure_numpy_condition_notations.py]: Not sure, [`futurize`](http://python-future.org/futurize.html) added that for me 😄 (I know, bad answer).

I don't think it really brings features per se. It just provides the module the Python 3 version of print. The only concrete advantages I can see:
-  if a contributor on Python 2 uses `print` the old way, they will get the error early, instead of waiting for the CI to crash.
- Running futurize again will not suggest the same change again

Comment (fpagnoux) [reply openfisca_core/scripts/measure_numpy_condition_notations.py]: Actually, I think to avoid confusion and to avoid inconsistency, let me just import all the __future__ in every module, on one line.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,14 @@
...
+  - Make code more Python3-like by backporting unicode litterals.
+  - With this backport, all strings are by default unicodes.
+  - The `u` prefix for strings should *not* be used anymore.
```
[the remark this answers is not in the document]
Comment (fpagnoux) [reply CHANGELOG.md]: If you can give me a shell command that does it, I can add it to Circle 🙂.

[the remark this answers is not in the document]
Comment (magopian) [reply CHANGELOG.md]: I know [black](https://github.com/ambv/black) does it automatically... but the change is a bit more involved... as you can see in https://github.com/openfisca/openfisca-france/pull/1060 (and specifically the commit https://github.com/openfisca/openfisca-france/pull/1060/commits/9dd758080c5efd20322093546852d1f1ad49e787 adding it, and the consequences: https://github.com/openfisca/openfisca-france/pull/1060/commits/6d8621b1d470e92a50de12ac92ff35d99803f228)

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,5 +1,14 @@
...
+  - Each new module must start by `from __future__ import unicode_literals` for the backport to be effective.
```
[the remark this answers is not in the document]
Comment (fpagnoux) [reply CHANGELOG.md]: Same, I'd happily take a shell command that makes sure that each `.py` file contains `from __future__ import unicode_literals, print_function, division, absolute_import`.

I guess I could do it in Python, but I'm sure there is a quick way in shell...

File: README.md [hunk 11]
```diff
@@ -6,7 +6,9 @@ This package contains the core features of OpenFisca, which are meant to be used
...
+
+Backward compatibility with Python 2.7 is for now guaranteed, but will be dropped from January 1st, 2019.
```
[the remark this answers is not in the document]
Comment (fpagnoux) [reply README.md]: Done

File: openfisca_core/columns.py [hunk 12]
```diff
@@ -32,6 +34,7 @@ def make_column_from_variable(variable):
         int: IntCol,
         float: FloatCol,
         str: StrCol,
+        bytes: StrCol,
```
[the remark this answers is not in the document]
Comment (fpagnoux) [reply openfisca_core/columns.py]: Actually, yes!

In Python `2`, by importing `str` from `builtins`, in this module `str` is now (roughly) Python 3's `str`.

But if in a country package I'm using the regular Python 2 `str` (which is the same than `bytes`) as a `value_type`, then I'll have a `KeyError` in `CONVERSION_MAP[variable.value_type]`.

So we need to add this to stay compatible with Python 2.

File: openfisca_core/columns.py [hunk 13]
```diff
@@ -1,5 +1,7 @@
 # -*- coding: utf-8 -*-
 
+from __future__ import unicode_literals
+from builtins import str
```
[the remark this answers is not in the document]
Comment (fpagnoux) [reply openfisca_core/columns.py]: Not a huge fan of `TODO` staying for months in the codebase...

I think when we'll drop Python 2, all the `from __future__`, `from builtins` and `from future` will be relatively easy to spot.  I'm adding a note about that to the [relevant issue](https://github.com/openfisca/openfisca-core/issues/693#issue-343550374) 🙂.

File: openfisca_core/scripts/find_placeholders.py [hunk 14]
```diff
@@ -1,5 +1,7 @@
 # -*- coding: utf-8 -*-
 
+from __future__ import print_function
```
[the remark this answers is not in the document]
Comment (fpagnoux) [reply openfisca_core/scripts/find_placeholders.py]: Switching to import all the __future__ in all python modules for consistency.

File: openfisca_web_api_preview/loader/variables.py (hunk 2 above)
Comment (fpagnoux) [reply openfisca_web_api_preview/loader/variables.py]: We haven't been using these decorators for a long time (FI for this kind of off-topic cleanups I try to explicit the intention in [commit titles](https://github.com/openfisca/openfisca-core/pull/702/commits/cd36929f13fc601d1a65a98bc1c4f77b5cf7fc5c))

Agreed, it's not related to the PR.

File: openfisca_core/variables.py [hunk 15]
```diff
@@ -358,6 +358,9 @@ def get_introspection_data(cls, tax_benefit_system):
             source_file_path = absolute_file_path.replace(tax_benefit_system.get_package_metadata()['location'], '')
         try:
             source_lines, start_line_number = inspect.getsourcelines(cls)
+            # Python 2 backward compatibility
+            if isinstance(source_lines[0], bytes):
+                source_lines = [source_line.decode('utf-8') for source_line in source_lines]
```
[the remark this answers is not in the document]
Comment (fpagnoux) [reply openfisca_core/variables.py]: Yep, tests caught this one. Switching to the syntax you suggest 👍

File: openfisca_web_api_preview/loader/tax_benefit_system.py (hunk 1 above)
Comment (fpagnoux) [reply openfisca_web_api_preview/loader/tax_benefit_system.py]: I guess `join` is flexible enough to handle both `bytes` and `str`:

`' '.join([u'X', b'Y'])` works in Python 2

File: openfisca_core/variables.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (fpagnoux) [reply openfisca_core/variables.py]: Arg, no, it actually doesn't work in Python 3, as it filters the values for which `source_line` is unicode! I guess we could use ternary instead, but that's becoming significantly heavier than the initial proposal. Reverting

**Comment by fpagnoux** [comment]
> FYI, black will automatically remove the u prefixes when the unicode_literals are used. It might make sense to adopt it and use it to automatically reformat the code? I gave it a go in openfisca/openfisca-france#1060 if you need an example of what it does.

I wish I had known about this before half-manually removing all the `u` 😞. Thanks for the tips!

> I saw a couple of changes that seem unrelated to the backport, not sure if they should be done in another PR?

Ideally, yes... I'd say that the size and complexity of the [riders](https://en.wikipedia.org/wiki/Rider_(legislation)) fall in the usual margin of tolerance, but I'm obviously biased by my self-interest.

File: CHANGELOG.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (fpagnoux) [reply CHANGELOG.md]: black seems very opinionated, so using it would be something to think about more thoroughly.

Or we can hope that shell wizzard @MattiSG drops by and finds a solution to this issue and the [other one](https://github.com/openfisca/openfisca-core/pull/702#discussion_r207300821) by the wave of a magic wand...

File: openfisca_web_api_preview/loader/tax_benefit_system.py (hunk 1 above)
Comment (magopian) [reply openfisca_web_api_preview/loader/tax_benefit_system.py]: I tried it in py2, and it didn't work: I believe it's the opposite of your example, and also you need to have a mix of chars that won't auto decode properly, like:

```python
u"\n".join(["é", u"é"])
```

This will fail, but I guess this specific use case never occured in real life. Wow, lucky, and yay for py3 and proper explicit unicode handling!
